textile/chat-features.textile

```diff
@@ -9,5 +9,148 @@ jump_to:
+
+Chat features are spread across multiple Realtime Channels. For the purpose of this section, a @contributor@ is a feature in Chat that is considered as part of the overall room status, as well as its lifecycle. A @contributor@ might share its realtime channel with another (e.g. messages and occupancy), or it might be entirely standalone (room reactions). In that sense, implementations of Room Lifecycle @MUST@ make no assumptions over which realtime channels are in use or being shared, and treat every @contributor@ as being entirely standalone.
+
+There are four room lifecycle operations: @ATTACH@, @DETACH@, @RELEASE@ and @RETRY@. @RETRY@ is an internal-only operation. In order to ensure the integrity of the chat room status, these operations are carried out atomically.
+
+h4(#rooms-lifecycle-operations). Room Lifecycle Operations
+
+* @(CHA-RL1)@ A room must be explicitly @ATTACHED@ to.
+** @(CHA-RL1a)@ @[Testable]@ If the room is already in the @ATTACHED@ status, then this operation is no-op.
+** @(CHA-RL1b)@ @[Testable]@ If the room is in the @RELEASING@ status, the operation shall be rejected with error code @102102@.
+** @(CHA-RL1c)@ @[Testable]@ If the room is in the @RELEASED@ status, the operation shall be rejected with error code @102103@.
+** @(CHA-RL1d)@ @[Testable]@ If a room lifecycle operation is already in progress, this operation shall wait until the current operation completes before continuing.
+** @(CHA-RL1e)@ @[Testable]@ Notwithstanding the above points, when the attachment operation begins, the room shall be transitioned to the @ATTACHING@ status.
+** @(CHA-RL1f)@ @[Testable]@ The underlying @contributors@ will have their Realtime Channels attached in sequence.
+** @(CHA-RL1g)@ When all @contributors@ Realtime Channels successfully attach, the operation is now complete and the room is considered attached.
+*** @(CHA-RL1g1)@ @[Testable]@ The room status shall be transitioned to @ATTACHED@.
+*** @(CHA-RL1g2)@ @[Testable]@ Any contributors to room that have a pending discontinuity event against them, must be notified of this fact, using the error that caused the original discontinuity.
+*** @(CHA-RL1g3)@ @[Testable]@ Any transient disconnect timeouts shall be cleared.
+** @(CHA-RL1h)@ If a one of the @contributors@ Realtime Channels fails to attach, then the operation has failed and must be rolled back.
+*** @(CHA-RL1h1)@ @[Testable]@ The @attach@ call must throw an @ErrorInfo@. The code for this error is determined by the feature that failed (see the @ErrorCodes@ enum in the JS SDK for more information).
+*** @(CHA-RL1h2)@ @[Testable]@ If the underlying Realtime Channel entered the @SUSPENDED@ state, then the room status will transition to @SUSPENDED@, using the Realtime Channel error as the @cause@ field of the @ErrorInfo@.
```

Comment:
This could be more explicit — does it mean "if the call to `contributor.attach()` fails, then check the contributor’s `state`. If it's SUSPENDED, then transition to SUSPENDED, using the contributor’s `errorReason` property as the `cause` field of the `error` of the emitted status change"?

Also, what `code` and `statusCode` should this error have?

Reply:
"the Realtime Channel error" is still a bit unclear — it could mean the error that the contributor `attach()` call failed with. Can we explicitly say to use the contributor’s `errorReason`?

Reply:
Dibe!

Reply:
Sorry, I’m not sure what you’ve changed — CHA-RL1h2 doesn't seem to have changed from bcb7390 to b4a495e?

Reply:
Is it meant to be using the same language as CHA-RL1h4 now uses — that is, to use the error thrown by channel `attach()`

Reply:
Sorry, have clarified to use same language as `CHA-RL1h4`